### PR TITLE
HTTP/2-based provider protocol

### LINKED ISSUE #362: Callbacks instead of event listeners?
according to the example https://github.com/argon/node-apn/blob/master/examples/sending-to-multiple-devices.js you are sending notifications to multiple devices but cannot tell when the sending of all devices is finished (of course using counters and other methods can be employed)

Would it not be better to use callbacks (err,result) paradigm or at least promises built in? So it would be best integrated with node-async for example.

This way we need to keep track of notifications that are returned in the event listeners argument.

### PR DESCRIPTION
**Pull request by argon**
In WWDC15 Session 720, a new provider protocol was introducted based on HTTP/2. This branch will implement the new support.

- [x] HTTP/2 Notification sending
- [x] Topic Support
- [x] Multiple Connections
  - [x] Round robin connections
- [x] Mock interface
- [x] `shutdown`
- [x] Error handling
  - [x] Wrap errors

I will move this work to separate PR's
- Rewrite documentation
  - Sample code
- Prepare API for new authentication method (JWT)

Fixes #362, #277, #240, #168

### CODE REVIEW

**Comment by jorisvervuurt** [comment]
FYI:
The HTTP/2 APNs provider API is now up-and running.
Documentation can be found here:
https://developer.apple.com/library/mac/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/Chapters/APNsProviderAPI.html#//apple_ref/doc/uid/TP40008194-CH101-SW1

Any ETA as to when we can expect a first version that supports it? :)

**Comment by argon** [comment]
I am still working on this. It's basically a complete rewrite though so it'll take a while.

**Comment by daohoangson** [comment]
Do you have a check list of issues to complete this PR? I would love to help to push this through asap.

**Comment by argon** [comment]
Not that the moment but I've been making good progress over the past week, I now have notifications sending but no queueing, multi-connection support or custom headers.

After the queueing and headers are done I'll write the documentation for the API and publish as a beta so people can start supporting the new version (2.0) in their code. The multi-connection support is not a small task but shouldn't take too long to implement. After that will be a number of necessary tweaks into the underlying library, including investigation into a possible memory leak.

**Comment by ericdolson** [comment]
Just wondering what kind of timeframe you are thinking on this? I may be interested in contributing if needed. Thanks!

**Comment by argon** [comment]
I have published an initial alpha version to npm, install with `npm install apn@alpha`. The documentation hasn't been updated yet but the interface should be familiar.

```javascript
"use strict";

const apn = require('apn');

let tokens = ["a9d0fd10e9cfd022a61cb08753f49c5a0b0dfb393697bf9f9d750a1003da19c7", ...]
let service = new apn.Connection({
  cert: "certificates/cert.pem",
  key: "certificates/key.pem",
  production: false,
});

let note = new apn.Notification();
note.alert = "This is a test";
note.topic = "io.apn.pushtest";

service.pushNotification(note, tokens).then( result => {
    console.log("sent:", result.sent.length);
    console.log("failed:", result.failed.length);
});
```

**Comment by argon** [comment]
Interface has been updated
```javascript
"use strict";

const apn = require('apn');

let tokens = [
  "a9d0fd10e9cfd022a61cb08753f49c5a0b0dfb393697bf9f9d750a1003da19c7",
  ...
].map( token => apn.token(token) );

let service = new apn.Provider({
  cert: "certificates/cert.pem",
  key: "certificates/key.pem",
  production: false,
});

let note = new apn.Notification();
note.alert = "This is a test";
note.topic = "io.apn.pushtest";

service.send(note, tokens).then( result => {
    console.log("sent:", result.sent.length);
    console.log("failed:", result.failed.length);
});
```

**Comment by argon** [comment]
@maccheck I missed that mention, thanks for the heads up. I'll remove this check.

**Comment by n3trino** [comment]
Any estimation on ETA?

**Comment by argon** [comment]
I'm going to merge to develop this week

—A

**Comment by flovilmart** [comment]
Will it be backwards compatible for node < 5?

File: lib/client.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
+"use strict";
+
+const Promise = require("bluebird");
```
Comment (flovilmart) [inline lib/client.js]: Promise are supported in the current node LTS (4.5), not sure if it's required to have an additional library like bluebird.

Comment (argon) [inline lib/client.js]: http://programmers.stackexchange.com/questions/278778/why-are-native-es6-promises-slower-and-more-memory-intensive-than-bluebird

**Comment by argon** [comment]
@flovilmart Tests are run against node 4 and it should continue to function.

Node 5 adds ALPN support which in theory is a requirement for HTTP/2 to work correctly however I don't think it applies to APNS

**Comment by flovilmart** [comment]
@argon thanks for the reply, I ran into troubles without ALPN support in the past with the http2 library with E2E tests with iOS push. Maybe that's been resolved since :)

File: lib/client.js (hunk 1 above)
Comment (flovilmart) [inline lib/client.js]: Yeah, I'm well aware of that, in the context I'm talking from (parse-server) we try to not embed those kind of optimizations and leave it for the users to 'opt-in' to reduce the dependencies surface. Kinda show stopper for me, but I may wrap my head around it.

Comment (argon) [inline lib/client.js]: What do you propose?

**Comment by argon** [comment]
I just ran my test script on node 4.5.0 and it functioned correctly with this branch.

File: lib/client.js (hunk 1 above)
Comment (argon) [inline lib/client.js]: @flovilmart how do you feel about it being a `devDependency` only? It makes some of the tests a little cleaner but could be foregone in the main library.

Comment (flovilmart) [inline lib/client.js]: That would be nice, with a conditional import, so if someone is providing bluebird in it's own project, that would use the external one. What do you think? I could very well do thé PR for that as it's more for me than you.

Comment (argon) [inline lib/client.js]: I'm already looking at this. I am fixing up the tests to not depend on bluebird and will put in a conditional import

Comment (flovilmart) [inline lib/client.js]: That's awesome thanks man! Highly appreciated. If you need anything, don't hesitate to ping :)

**Pull request merged**